File: leetcode/Python/minimum-number-of-refueling-stops.py
class Solution:
    def minRefuelStops(self, target: int, startFuel: int, stations: List[List[int]]) -> int:
        visit, pq = [0 for i in range(len(stations))], [(0,-startFuel,-1,0,startFuel)]
        info,f = None,0
        while pq:
            info = heapq.heappop(pq)
            if info[3]+info[4] >= target:
                return info[0]
            if info[2]>=0 and visit[info[2]]>info[3]+info[4]:
                continue
            for i in range(info[2]+1,len(stations)):
                if info[3]+info[4]>=stations[i][0]:
                    f = stations[i][1]+info[3]+info[4]
                    if visit[i]<f:
                        visit[i]=f
                        heapq.heappush(pq,(info[0]+1,-f,i,stations[i][0],f-stations[i][0]))
                else:
                    break
        return -1

File: leetcode/C++/minimum-number-of-refueling-stops.cpp
struct INFO{
    int recharge;
    int idx;
    int fuel;
    int location;
    bool operator()(INFO &a, INFO &b){
        if(a.recharge == b.recharge) return a.location+a.fuel < b.location+b.fuel;
        return a.recharge>b.recharge;
    }
};

class Solution {
public:
    int minRefuelStops(int target, int startFuel, vector<vector<int>>& stations) {
        INFO info = {0,-1,startFuel,0};
        vector<int> visit(stations.size(),0);
        priority_queue<INFO,vector<INFO>,INFO> pq;
        pq.push({0,-1,startFuel,0});
        while(!pq.empty()){
            info = pq.top(); pq.pop();
            if(info.idx>=0 && visit[info.idx]>info.location+info.fuel) continue;
            if(info.location+info.fuel >= target) return info.recharge;
            for(int i=info.idx+1; i<stations.size(); ++i){
                if(info.location+info.fuel >= stations[i][0]){
                    if(visit[i]<info.fuel+info.location+stations[i][1]) {
                        visit[i]=info.fuel+info.location+stations[i][1];
                        pq.push({info.recharge+1,i,info.fuel-stations[i][0]+info.location+stations[i][1],stations[i][0]});
                    }
                   
                }
                else break;
            }
        }
        return -1;
    }
};